course_content/solutions/numpy_exercise_10.ipynb

```diff
@@ -0,0 +1,305 @@
+     "metadata": {
+      "slideshow": {
+       "slide_type": "-"
+      }
+     },
+     "source": [
+      "#### Part 4\n",
+      "\n",
+      "Verify that your function is correct by using the arrays created in #1 as input to ``trapz``. Your answer should be a close approximation of $\\int_0^3 x^2$ which is $9$."
+     ]
+    },
+    {
+     "cell_type": "code",
+     "collapsed": false,
+     "input": [
+      "trapz(x, y)"
+     ],
+     "language": "python",
+     "metadata": {
+      "slideshow": {
+       "slide_type": "skip"
+      }
+     },
+     "outputs": [
+      {
+       "metadata": {},
+       "output_type": "pyout",
+       "prompt_number": 4,
+       "text": [
+        "9.0555555555555554"
+       ]
+      }
+     ],
+     "prompt_number": 4
+    },
+    {
+     "cell_type": "markdown",
+     "metadata": {
+      "slideshow": {
+       "slide_type": "-"
+      }
+     },
+     "source": [
+      "#### Part 5 (extension)\n",
+      "\n",
+      "``numpy`` and ``scipy.integrate`` provides many common integration schemes.  Find the documentation for NumPy's own version of the trapezoidal integration scheme and check its result with your own:"
+     ]
+    },
+    {
+     "cell_type": "code",
+     "collapsed": false,
+     "input": [
+      "print np.trapz(y, x)"
+     ],
+     "language": "python",
+     "metadata": {
+      "slideshow": {
+       "slide_type": "skip"
+      }
+     },
+     "outputs": [
+      {
+       "output_type": "stream",
+       "stream": "stdout",
+       "text": [
+        "9.05555555556\n"
+       ]
+      }
+     ],
+     "prompt_number": 5
+    },
+    {
+     "cell_type": "markdown",
+     "metadata": {
+      "slideshow": {
+       "slide_type": "-"
+      }
+     },
+     "source": [
+      "#### Part 6 (extension)\n",
+      "\n",
+      "Write a function  `trapzf(f, a, b, npts=100)` that accepts a function `f`, the endpoints `a` and `b` and the number of samples to take `npts`.  Sample the function uniformly at these\n",
+      "points and return the value of the integral.\n",
+      "\n",
+      "Use the trapzf function to identify the minimum number of sampling points needed to approximate the integral $\\int_0^3 x^2$ with an absolute error of $<=0.0001$. (A loop is necessary here)"
+     ]
+    },
+    {
+     "cell_type": "code",
+     "collapsed": false,
+     "input": [
+      "def trapzf(f, a, b, npts=100):\n",
+      "    x = np.linspace(a, b, npts)\n",
+      "    y = f(x)\n",
+      "    return trapz(x, y)\n",
+      "\n",
+      "def x_squared(x):\n",
+      "    return x ** 2\n",
+      "\n",
+      "abs_err = 1.0\n",
+      "n_samples = 0\n",
+      "expected = 9\n",
+      "while abs_err > 0.0001:\n",
+      "    n_samples += 1\n",
+      "    integral = trapzf(x_squared, 0, 3, npts=n_samples)\n",
+      "    abs_err = np.abs(integral - 9)\n",
+      "\n",
+      "print 'Minimum samples for absolute error less than or equal to 0.0001:', n_samples\n",
```

Comment:
`print('Minimum samples for absolute error less than or equal to 0.0001:', n_samples)`